File: src/main/scala/committee/nova/skillful/impl/skill/instance/SkillInstance.scala
package committee.nova.skillful.impl.skill.instance

import committee.nova.skillful.api.skill.ISkill
import committee.nova.skillful.event.impl.{SkillLevelEvent, SkillXpEvent}
import committee.nova.skillful.implicits.Implicits.PlayerEntityImplicit
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.nbt.CompoundNBT
import net.minecraft.util.math.MathHelper
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.common.util.INBTSerializable

class SkillInstance(val skill: ISkill) extends INBTSerializable[CompoundNBT] {
  private var currentXp: Int = 0

  private var currentLevel: Int = 0

  def getSkill: ISkill = skill

  def getCurrentXp: Int = currentXp

  def getCurrentLevel: Int = currentLevel

  def isAcquired: Boolean = getCurrentLevel > 0

  def isClueless: Boolean = !isAcquired && currentXp == 0

  def isCompleted: Boolean = getCurrentLevel >= getSkill.getMaxLevel

  def addXp(player: ServerPlayerEntity, xp: Int): Unit = {
    if (player.isFake) return
    val event = new SkillXpEvent.Pre(player, this, xp)
    if (MinecraftForge.EVENT_BUS.post(event)) return
    _addXp(player, event.getAmount)
  }

  private def _addXp(player: ServerPlayerEntity, xp: Int): Unit = {
    if (player.isFake) return
    if (xp == 0) return
    if (xp < 0) {
      _reduceXp(player, -xp)
      return
    }
    if (isCompleted) return
    currentXp += xp min (Int.MaxValue - currentXp)
    while (currentXp >= skill.getLevelRequiredXp(currentLevel)) {
      currentXp -= skill.getLevelRequiredXp(currentLevel)
      currentLevel += 1
      if (currentLevel > skill.getMaxLevel) complete()
      else MinecraftForge.EVENT_BUS.post(new SkillLevelEvent.Up(player, this, currentLevel))
    }
    MinecraftForge.EVENT_BUS.post(new SkillXpEvent.Post(player, this, xp))
    player.syncSkills()
  }

  def reduceXp(player: ServerPlayerEntity, xp: Int): Unit = {
    if (player.isFake) return
    val event = new SkillXpEvent.Pre(player, this, -xp)
    if (MinecraftForge.EVENT_BUS.post(event)) return
    _reduceXp(player, -event.getAmount)
  }

  private def _reduceXp(player: ServerPlayerEntity, xp: Int): Unit = {
    if (player.isFake) return
    if (xp == 0) return
    if (xp < 0) {
      _addXp(player, -xp)
      return
    }
    if (isClueless) return
    currentXp -= xp
    while (currentXp < 0) {
      currentLevel -= 1
      currentXp += skill.getLevelRequiredXp(currentLevel)
      if (currentLevel < 0) makeClueless()
      else MinecraftForge.EVENT_BUS.post(new SkillLevelEvent.Down(player, this, currentLevel))
    }
    MinecraftForge.EVENT_BUS.post(new SkillXpEvent.Post(player, this, -xp))
    player.syncSkills()
  }

  def clear(player: ServerPlayerEntity): Unit = reduceXp(player, Int.MaxValue)

  def cheat(player: ServerPlayerEntity): Unit = addXp(player, Int.MaxValue)

  def changeLevel(player: ServerPlayerEntity, lvl: Int): Unit = {
    if (player.isFake) return
    val real = MathHelper.clamp(lvl, 0, skill.getMaxLevel)
    val old = getCurrentLevel
    currentLevel = real
    currentXp = 0
    if (old == real) return
    MinecraftForge.EVENT_BUS.post(if (real > old) new SkillLevelEvent.Up(player, this, real) else new SkillLevelEvent.Down(player, this, real))
    player.syncSkills()
  }

  private def makeClueless(): Unit = {
    currentXp = 0
    currentLevel = 0
  }

  private def complete(): Unit = {
    currentXp = 0
    currentLevel = skill.getMaxLevel
  }

  override def serializeNBT(): CompoundNBT = {
    val tag = new CompoundNBT
    tag.putString("skill", skill.getId.toString)
    tag.putInt("level", currentLevel)
    tag.putInt("xp", currentXp)
    tag
  }

  override def deserializeNBT(tag: CompoundNBT): Unit = {
    currentLevel = tag.getInt("level")
    currentXp = tag.getInt("xp")
  }

  override def equals(obj: Any): Boolean = obj match {
    case that: SkillInstance => that.skill.equals(skill)
    case _ => false
  }

  override def hashCode(): Int = skill.hashCode()
}

File: src/main/java/committee/nova/skillful/common/skill/SkillInstance.java
package committee.nova.skillful.common.skill;

import committee.nova.skillful.common.event.impl.SkillLevelEvent;
import committee.nova.skillful.common.util.Utilities;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraftforge.common.MinecraftForge;

public class SkillInstance implements ISkillRecord {
    private final ISkillType skill;
    private long level;
    private long xp;

    private SkillInstance(ISkillType skill) {
        this.skill = skill;
    }

    public static SkillInstance of(ISkillType skill) {
        return new SkillInstance(skill);
    }

    public ISkillType getSkill() {
        return skill;
    }

    public boolean acquired() {
        return level > 0;
    }

    public boolean clueless() {
        return !acquired() && xp == 0;
    }

    public boolean mastered() {
        return level >= skill.getMaxLevel();
    }

    public void changeXp(ServerPlayer player, long change) {
        if (Utilities.isPlayerFake(player)) return;
        if (change >= 0) _addXp(player, change);
        else _reduceXp(player, -change);
    }

    public void changeToLevel(ServerPlayer player, long target) {
        if (Utilities.isPlayerFake(player)) return;
        final long level0 = level;
        level = Math.max(0, Math.min(target, skill.getMaxLevel()));
        if (level0 == level) return;
        xp = 0;
        MinecraftForge.EVENT_BUS.post(new SkillLevelEvent(player, this, level0, level));
        skill.onLevelChange(player, this, level0, level);
        syncToClient(player);
    }

    private void _addXp(ServerPlayer player, long change) {
        if (change == 0) return;
        if (change < 0) {
            _reduceXp(player, -change);
            return;
        }
        if (mastered()) return;
        xp += Math.min(change, Long.MAX_VALUE - xp);
        final long level0 = level;
        while (true) {
            final long required = skill.getLevelRequiredXp(player, level);
            if (xp < required) break;
            level++;
            if (level > skill.getMaxLevel()) {
                complete();
                break;
            }
            xp -= required;
        }
        if (level0 != level) {
            skill.onLevelChange(player, this, level0, level);
            MinecraftForge.EVENT_BUS.post(new SkillLevelEvent(player, this, level0, level));
        }
        syncToClient(player);
    }

    private void _reduceXp(ServerPlayer player, long change) {
        if (change == 0) return;
        if (change < 0) {
            _addXp(player, -change);
            return;
        }
        final long level0 = level;
        if (clueless()) return;
        xp -= change;
        while (xp < 0) {
            level--;
            if (level < 0) {
                clear();
                break;
            }
            xp += skill.getLevelRequiredXp(player, level);
        }
        if (level0 != level) {
            skill.onLevelChange(player, this, level0, level);
            MinecraftForge.EVENT_BUS.post(new SkillLevelEvent(player, this, level0, level));
        }
        syncToClient(player);
    }

    private void complete() {
        level = skill.getMaxLevel();
        xp = 0;
    }

    private void clear() {
        level = 0;
        xp = 0;
    }

    @Override
    public ResourceLocation getId() {
        return skill.getId();
    }

    @Override
    public long getLevel() {
        return level;
    }

    @Override
    public long getXp() {
        return xp;
    }

    @Override
    public void setLevel(long level) {
        this.level = level;
    }

    @Override
    public void setXp(long xp) {
        this.xp = xp;
    }
}
